homeassistant/components/denonavr/config_flow.py

```diff
@@ -0,0 +1,255 @@
+
+    async def async_step_user(self, user_input=None):
+        """Handle a flow initialized by the user."""
+        errors = {}
+        if user_input is not None:
+            # check if IP address is set manually
+            host = user_input.get(CONF_HOST)
+            if host:
+                self.host = host
+                return await self.async_step_connect()
+
+            # discovery using denonavr library
+            self.d_receivers = await self.hass.async_add_executor_job(denonavr.discover)
+            # More than one receiver could be discovered by that method
+            if len(self.d_receivers) == 1:
+                self.host = self.d_receivers[0]["host"]
+                return await self.async_step_connect()
+            if len(self.d_receivers) > 1:
+                # show selection form
+                return await self.async_step_select()
+
+            errors["base"] = "discovery_error"
+
+        return self.async_show_form(
+            step_id="user", data_schema=CONFIG_SCHEMA, errors=errors
+        )
+
+    async def async_step_select(self, user_input=None):
+        """Handle multiple receivers found."""
+        errors = {}
+        if user_input is not None:
+            self.host = user_input["select_host"]
+            return await self.async_step_connect()
+
+        select_scheme = vol.Schema(
+            {
+                vol.Required("select_host"): vol.In(
+                    [d_receiver["host"] for d_receiver in self.d_receivers]
+                )
+            }
+        )
+
+        return self.async_show_form(
+            step_id="select", data_schema=select_scheme, errors=errors
+        )
+
+    async def async_step_confirm(self, user_input=None):
+        """Allow the user to confirm adding the device."""
+        if user_input is not None:
+            return await self.async_step_connect()
+
+        return self.async_show_form(step_id="confirm")
+
+    async def async_step_connect(self, user_input=None):
+        """Connect to the receiver."""
+        connect_denonavr = ConnectDenonAVR(
+            self.hass,
+            self.host,
+            self.timeout,
+            self.show_all_sources,
+            self.zone2,
+            self.zone3,
+        )
+        if not await connect_denonavr.async_connect_receiver():
+            return self.async_abort(reason="connection_error")
+        receiver = connect_denonavr.receiver
+
+        mac_address = await self.async_get_mac(self.host)
+
+        if not self.serial_number:
+            self.serial_number = receiver.serial_number
+        if not self.model_name:
+            self.model_name = (receiver.model_name).replace("*", "")
+
+        if self.serial_number is not None:
+            unique_id = self.construct_unique_id(self.model_name, self.serial_number)
+        else:
+            _LOGGER.error(
+                "Could not get serial number of host %s, "
+                "using the mac_address as identification",
```

Comment:
Can we assume that a device will use the same unique id, serial or mac, regardless if it's set up via the user or via discovery?

What devices do not provide a serial number? How often does that happen?

Reply:
Yes it will in principle always use the serial number regardless of the way you set up (manually or discovery).

If the serial is not available on discovery the manually setup will also not have the serial number available because the denonavr lib uses the same xml file to get the serial as the discovery component does.

However we can of course not 100% garantuee that this is the case, but the whole using mac_address is just a backup method that issues errors so that users can report the error but in the mean time be able to use the integration. We schould use the serial number and if that is not available users schould report this error and it schould be fixed.

Reply:
Mac address is also not always available so I don't think it's a good fall back.

I think it would be better to fall back to no unique id, when serial number is missing, and then just check the existing config entry host addresses to make sure we don't set up the same host again. With no unique id, we return `None` for `device_info` and `unique_id` in the entity.

Reply:
All devices that are tested and known to me always provide serial number.
However since Denon/Marantz have the really enoying habit of implementing there API just a litlle bit diffrent for each of there models I can not 100% garantuee that it will work for all models.

(like sound modes are reported with extra spaces or just spelled diffrently in the denon/marantz API, really anoying... for example 'MULTI CH STEREO', 'MULTI_CH_STEREO', 'MCH STEREO')

Reply:
Yeah, I think that's another reason why we shouldn't make this too complicated. If the known devices report serial number, let's go with that and fall back to no unique id otherwise.

Reply:
alright, will change that tonight

Reply:
I removed the mac_adress fallback and allowed the integration to setup withouth an unique_id.